File: rogue/game/entities.py
"""Base entity classes shared by the player, monsters, and items."""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, List, Optional, Tuple

from .constants import (
    PLAYER_START_HP, PLAYER_START_STR, PLAYER_START_AC,
    PLAYER_EXP_TABLE, HUNGER_FULL,
    CONFUSED_TURNS, BLIND_TURNS, HASTED_TURNS, POISONED_TURNS,
    FROZEN_TURNS, HALLUC_TURNS,
    MAX_INVENTORY, WHITE,
)

if TYPE_CHECKING:
    from .items import Item


# ---------------------------------------------------------------------------
# Entity  (anything that occupies a tile)
# ---------------------------------------------------------------------------

class Entity:
    """A map object with a position, display character and colour."""

    def __init__(
        self,
        x: int, y: int,
        char: str,
        color: Tuple[int, int, int],
        name: str,
    ) -> None:
        self.x = x
        self.y = y
        self.char = char
        self.color = color
        self.name = name

    @property
    def pos(self) -> Tuple[int, int]:
        return (self.x, self.y)

    def move_to(self, x: int, y: int) -> None:
        self.x = x
        self.y = y


# ---------------------------------------------------------------------------
# Actor  (entity with hit-points and combat stats)
# ---------------------------------------------------------------------------

class Actor(Entity):
    """Anything that can fight: player or monster."""

    def __init__(
        self,
        x: int, y: int,
        char: str,
        color: Tuple[int, int, int],
        name: str,
        max_hp: int,
        attack_dice: Tuple[int, int],   # (num_dice, sides) e.g. (2, 6) = 2d6
        defense: int,                    # raw armor class (lower = better)
        xp_value: int = 0,
    ) -> None:
        super().__init__(x, y, char, color, name)
        self.max_hp = max_hp
        self.hp = max_hp
        self.attack_dice = attack_dice  # (n, sides)
        self.defense = defense
        self.xp_value = xp_value

        # Status effects  (turns remaining, 0 = inactive)
        self.confused  = 0
        self.blinded   = 0
        self.hasted    = 0
        self.poisoned  = 0
        self.frozen    = 0
        self.sleeping  = 0

    @property
    def alive(self) -> bool:
        return self.hp > 0

    def take_damage(self, amount: int) -> int:
        """Reduce HP and return actual damage taken."""
        amount = max(1, amount)
        self.hp = max(0, self.hp - amount)
        return amount

    def heal(self, amount: int) -> int:
        """Restore HP up to max; return actual amount healed."""
        healed = min(amount, self.max_hp - self.hp)
        self.hp += healed
        return healed

    def roll_attack(self) -> int:
        n, sides = self.attack_dice
        return sum(random.randint(1, sides) for _ in range(n))

    def tick_effects(self) -> List[str]:
        """Decrement status timers; return list of expiry messages."""
        msgs = []
        for attr, label in (
            ("confused", "You are no longer confused."),
            ("blinded",  "Your vision clears."),
            ("hasted",   "You slow down."),
            ("poisoned", "The poison wears off."),
            ("frozen",   "You can move again."),
            ("sleeping", ""),
        ):
            val = getattr(self, attr)
            if val > 0:
                setattr(self, attr, val - 1)
                if getattr(self, attr) == 0 and label:
                    msgs.append(label)
        return msgs


# ---------------------------------------------------------------------------
# Player
# ---------------------------------------------------------------------------

class Player(Actor):
    """The player character."""

    def __init__(self, x: int, y: int) -> None:
        super().__init__(
            x=x, y=y,
            char='@',
            color=WHITE,
            name="you",
            max_hp=PLAYER_START_HP,
            attack_dice=(1, 4),      # bare-hands damage
            defense=PLAYER_START_AC,
            xp_value=0,
        )
        self.str_base = PLAYER_START_STR
        self.str_cur  = PLAYER_START_STR
        self.exp_level = 1
        self.exp_pts   = 0
        self.gold      = 0
        self.hunger    = HUNGER_FULL

        # Inventory: list of Item objects (max MAX_INVENTORY)
        self.inventory: List["Item"] = []
        # Equipment slots
        self.weapon: Optional["Item"] = None   # wielded weapon
        self.armor:  Optional["Item"] = None   # worn armor

        # Effective AC (recalculated from armor)
        self._armor_ac: int = 0

        # Has the Amulet of Yendor?
        self.has_amulet = False

        # Hallucination / see-invisible
        self.hallucinating = 0
        self.see_invisible = 0

    # -- AC ----------------------------------------------------------

    @property
    def effective_ac(self) -> int:
        """Lower is better.  Armour subtracts from base AC."""
        return PLAYER_START_AC - self._armor_ac

    def recalc_ac(self) -> None:
        base = 0
        if self.armor:
            base = self.armor.ac_bonus
        self._armor_ac = base

    # -- STR modifiers -----------------------------------------------

    @property
    def str_damage_bonus(self) -> int:
        if self.str_cur >= 18: return 6
        if self.str_cur >= 16: return 4
        if self.str_cur >= 13: return 2
        if self.str_cur >= 10: return 0
        if self.str_cur >= 7:  return -1
        return -2

    # -- Attack ------------------------------------------------------

    def roll_attack(self) -> int:
        if self.weapon:
            n, sides = self.weapon.damage_dice
        else:
            n, sides = self.attack_dice  # bare hands
        dmg = sum(random.randint(1, sides) for _ in range(n))
        if self.weapon:
            dmg += self.weapon.damage_bonus
        dmg += self.str_damage_bonus
        return max(1, dmg)

    # -- Experience --------------------------------------------------

    def gain_exp(self, amount: int) -> Optional[str]:
        """Add exp; return level-up message if applicable, else None."""
        self.exp_pts += amount
        msg = None
        while (self.exp_level < len(PLAYER_EXP_TABLE) - 1 and
               self.exp_pts >= PLAYER_EXP_TABLE[self.exp_level]):
            self.exp_level += 1
            hp_gain = random.randint(3, 10)
            self.max_hp += hp_gain
            self.hp += hp_gain
            msg = f"Welcome to level {self.exp_level}!  You feel stronger."
        return msg

    # -- Inventory ---------------------------------------------------

    def add_item(self, item: "Item") -> Optional[str]:
        """Add item to inventory. Returns slot letter, or None if full."""
        if len(self.inventory) >= MAX_INVENTORY:
            return None
        self.inventory.append(item)
        return chr(ord('a') + len(self.inventory) - 1)

    def remove_item(self, item: "Item") -> None:
        if item in self.inventory:
            self.inventory.remove(item)
            # If it was equipped, unequip
            if self.weapon is item:
                self.weapon = None
            if self.armor is item:
                self.armor = None
                self.recalc_ac()

    def item_slot(self, item: "Item") -> str:
        """Return inventory letter for this item."""
        idx = self.inventory.index(item)
        return chr(ord('a') + idx)

    # -- Hunger ------------------------------------------------------

    def tick_hunger(self) -> Optional[str]:
        """Consume one hunger unit.  Returns warning message if needed."""
        self.hunger -= 1
        if self.hunger == 300:
            return "You are starting to feel hungry."
        if self.hunger == 150:
            return "You are feeling weak!"
        if self.hunger == 20:
            return "You are about to faint from hunger!"
        if self.hunger <= 0:
            self.hunger = 0
            # Starvation damage every 10 turns
            return None   # handled by caller
        return None

    def eat(self, nutrition: int) -> None:
        self.hunger = min(HUNGER_FULL, self.hunger + nutrition)

    # -- Tick --------------------------------------------------------

    def tick_effects(self) -> List[str]:
        msgs = super().tick_effects()
        if self.hallucinating > 0:
            self.hallucinating -= 1
            if self.hallucinating == 0:
                msgs.append("Everything looks normal again.")
        if self.see_invisible > 0:
            self.see_invisible -= 1
        return msgs

File: rogue-app/src/engine/entities.ts
// Base entity classes shared by the player, monsters, and items.

import {
  PLAYER_START_HP,
  PLAYER_START_STR,
  PLAYER_START_AC,
  PLAYER_EXP_TABLE,
  HUNGER_FULL,
  MAX_INVENTORY,
  WHITE,
  RGB,
  strPlus,
  addDam,
} from './constants';
import { rng } from './rng';
import type { Item, Ring } from './items';

// ---------------------------------------------------------------------------
// Entity  (anything that occupies a tile)
// ---------------------------------------------------------------------------

export class Entity {
  x: number;
  y: number;
  char: string;
  color: RGB;
  name: string;

  constructor(x: number, y: number, char: string, color: RGB, name: string) {
    this.x = x;
    this.y = y;
    this.char = char;
    this.color = color;
    this.name = name;
  }

  get pos(): [number, number] {
    return [this.x, this.y];
  }

  moveTo(x: number, y: number): void {
    this.x = x;
    this.y = y;
  }
}

// ---------------------------------------------------------------------------
// Actor  (entity with hit-points and combat stats)
// ---------------------------------------------------------------------------

export class Actor extends Entity {
  maxHp: number;
  hp: number;
  attackDice: [number, number]; // (n, sides)
  defense: number;
  xpValue: number;

  // Status effects (turns remaining, 0 = inactive)
  confused = 0;
  blinded = 0;
  hasted = 0;
  poisoned = 0;
  frozen = 0;
  sleeping = 0;

  constructor(
    x: number,
    y: number,
    char: string,
    color: RGB,
    name: string,
    maxHp: number,
    attackDice: [number, number],
    defense: number,
    xpValue = 0,
  ) {
    super(x, y, char, color, name);
    this.maxHp = maxHp;
    this.hp = maxHp;
    this.attackDice = attackDice;
    this.defense = defense;
    this.xpValue = xpValue;
  }

  get alive(): boolean {
    return this.hp > 0;
  }

  takeDamage(amount: number): number {
    amount = Math.max(1, amount);
    this.hp = Math.max(0, this.hp - amount);
    return amount;
  }

  heal(amount: number): number {
    const healed = Math.min(amount, this.maxHp - this.hp);
    this.hp += healed;
    return healed;
  }

  rollAttack(): number {
    const [n, sides] = this.attackDice;
    let total = 0;
    for (let i = 0; i < n; i++) total += rng.randint(1, sides);
    return total;
  }

  tickEffects(): string[] {
    const msgs: string[] = [];
    const specs: [keyof Actor, string][] = [
      ['confused', 'You are no longer confused.'],
      ['blinded', 'Your vision clears.'],
      ['hasted', 'You slow down.'],
      ['poisoned', 'The poison wears off.'],
      ['frozen', 'You can move again.'],
      ['sleeping', ''],
    ];
    for (const [attr, label] of specs) {
      const val = this[attr] as unknown as number;
      if (val > 0) {
        (this[attr] as unknown as number) = val - 1;
        if ((this[attr] as unknown as number) === 0 && label) msgs.push(label);
      }
    }
    return msgs;
  }
}

// ---------------------------------------------------------------------------
// Player
// ---------------------------------------------------------------------------

export class Player extends Actor {
  strBase: number;
  strCur: number;
  expLevel: number;
  expPts: number;
  gold: number;
  hunger: number;

  inventory: Item[] = [];
  weapon: Item | null = null;
  armor: Item | null = null;
  rings: Ring[] = []; // worn rings (max 2)

  private armorAc = 0;

  hasAmulet = false;
  hallucinating = 0;
  seeInvisible = 0;
  levitating = 0;
  confusingTouch = false; // scroll of monster confusion — next hit confuses
  private regenCounter = 0;
  private digestion = 0;

  constructor(x: number, y: number) {
    super(x, y, '@', WHITE, 'you', PLAYER_START_HP, [1, 4], PLAYER_START_AC, 0);
    this.strBase = PLAYER_START_STR;
    this.strCur = PLAYER_START_STR;
    this.expLevel = 1;
    this.expPts = 0;
    this.gold = 0;
    this.hunger = HUNGER_FULL;
  }

  // -- Rings -------------------------------------------------------

  hasRing(key: string): boolean {
    return this.rings.some((r) => r.effectKey === key);
  }

  /** Summed magnitude of all worn rings of a kind (protection, add_str, …). */
  ringBonus(key: string): number {
    return this.rings.filter((r) => r.effectKey === key).reduce((s, r) => s + (r.bonus || 0), 0);
  }

  get effectiveStr(): number {
    return this.strCur + this.ringBonus('add_str');
  }

  get canSeeInvisible(): boolean {
    return this.seeInvisible > 0 || this.hasRing('see_invisible');
  }

  /** Lower strength unless a ring of sustain strength prevents it. */
  reduceStr(n: number): boolean {
    if (this.hasRing('sustain_str')) return false;
    this.strCur = Math.max(1, this.strCur - n);
    return true;
  }

  // -- AC ----------------------------------------------------------

  get effectiveAc(): number {
    return PLAYER_START_AC - this.armorAc - this.ringBonus('protection');
  }

  recalcAc(): void {
    this.armorAc = this.armor ? this.armor.acBonus : 0;
  }

  // -- STR modifiers (Rogue 5.4.4 str_plus / add_dam) --------------

  /** To-hit bonus: strength + weapon enchantment + ring of dexterity. */
  get toHitBonus(): number {
    const weaponPlus = this.weapon ? (this.weapon as unknown as { enchant?: number }).enchant ?? 0 : 0;
    return strPlus(this.effectiveStr) + weaponPlus + this.ringBonus('dexterity');
  }

  // -- Attack ------------------------------------------------------

  rollAttack(): number {
    let n: number;
    let sides: number;
    if (this.weapon) [n, sides] = this.weapon.damageDice;
    else [n, sides] = this.attackDice;
    let dmg = 0;
    for (let i = 0; i < n; i++) dmg += rng.randint(1, sides);
    if (this.weapon) dmg += this.weapon.damageBonus;
    dmg += addDam(this.effectiveStr) + this.ringBonus('increase_damage');
    return Math.max(1, dmg);
  }

  // -- Experience --------------------------------------------------

  gainExp(amount: number): string | null {
    this.expPts += amount;
    let msg: string | null = null;
    while (this.expLevel < PLAYER_EXP_TABLE.length - 1 && this.expPts >= PLAYER_EXP_TABLE[this.expLevel]) {
      this.expLevel += 1;
      const hpGain = rng.randint(3, 10);
      this.maxHp += hpGain;
      this.hp += hpGain;
      msg = `Welcome to level ${this.expLevel}!  You feel stronger.`;
    }
    return msg;
  }

  // -- Inventory ---------------------------------------------------

  addItem(item: Item): string | null {
    if (this.inventory.length >= MAX_INVENTORY) return null;
    this.inventory.push(item);
    return String.fromCharCode('a'.charCodeAt(0) + this.inventory.length - 1);
  }

  removeItem(item: Item): void {
    const idx = this.inventory.indexOf(item);
    if (idx !== -1) {
      this.inventory.splice(idx, 1);
      if (this.weapon === item) this.weapon = null;
      if (this.armor === item) {
        this.armor = null;
        this.recalcAc();
      }
      const ri = this.rings.indexOf(item as unknown as Ring);
      if (ri !== -1) this.rings.splice(ri, 1);
    }
  }

  itemSlot(item: Item): string {
    const idx = this.inventory.indexOf(item);
    return String.fromCharCode('a'.charCodeAt(0) + idx);
  }

  // -- Hunger ------------------------------------------------------

  tickHunger(): string | null {
    this.hunger -= 1;
    // Worn rings speed digestion; a ring of slow digestion offsets the cost.
    if (this.rings.length > 0) {
      const cost = this.rings.filter((r) => r.effectKey !== 'slow_digestion' && r.effectKey !== 'adornment').length;
      this.digestion += cost - (this.hasRing('slow_digestion') ? 1 : 0);
      while (this.digestion >= 3) {
        this.hunger -= 1;
        this.digestion -= 3;
      }
      while (this.digestion <= -3) {
        this.hunger += 1;
        this.digestion += 3;
      }
    }
    if (this.hunger === 300) return 'You are starting to feel hungry.';
    if (this.hunger === 150) return 'You are feeling weak!';
    if (this.hunger === 20) return 'You are about to faint from hunger!';
    if (this.hunger <= 0) {
      this.hunger = 0;
      return null; // starvation damage handled by caller
    }
    return null;
  }

  eat(nutrition: number): void {
    this.hunger = Math.min(HUNGER_FULL, this.hunger + nutrition);
  }

  // -- Natural healing (Rogue daemon.c doctor) ---------------------

  /**
   * Regenerate HP over time. Below level 8 the hero heals 1 HP every
   * (21 - 2*level) turns; from level 8 on, every 3 turns for rnd(level-7)+1.
   */
  regen(): void {
    if (this.hp >= this.maxHp) {
      this.regenCounter = 0;
      return;
    }
    this.regenCounter += 1;
    let interval = this.expLevel < 8 ? Math.max(1, 21 - this.expLevel * 2) : 3;
    if (this.hasRing('regeneration')) interval = Math.max(1, Math.floor(interval / 2));
    if (this.regenCounter >= interval) {
      const amount = this.expLevel < 8 ? 1 : rng.randint(1, Math.max(1, this.expLevel - 7));
      this.heal(amount);
      this.regenCounter = 0;
    }
  }

  // -- Tick --------------------------------------------------------

  tickEffects(): string[] {
    const msgs = super.tickEffects();
    if (this.hallucinating > 0) {
      this.hallucinating -= 1;
      if (this.hallucinating === 0) msgs.push('Everything looks normal again.');
    }
    if (this.seeInvisible > 0) this.seeInvisible -= 1;
    if (this.levitating > 0) {
      this.levitating -= 1;
      if (this.levitating === 0) msgs.push('You float gently back to the ground.');
    }
    return msgs;
  }
}
